bazel/EnvoyMobileTestSuite.kt

```diff
@@ -0,0 +1,54 @@
+package io.envoyproxy.envoymobile.bazel
+
+import junit.framework.JUnit4TestAdapter
+import junit.framework.TestSuite
+import org.junit.runner.RunWith
+import java.io.File
+import java.net.URLClassLoader
+import java.util.zip.ZipFile
+
+/**
+ * This is class is taken from https://stackoverflow.com/questions/46365464/how-to-run-all-tests-in-bazel-from-a-single-java-test-rule
+ *
+ * Translated to Kotlin and slightly modified
+ *
+ */
+@RunWith(org.junit.runners.AllTests::class)
+object EnvoyMobileTestSuite {
+  private const val CLASS_SUFFIX = ".class"
+
+  @JvmStatic
+  fun suite(): TestSuite {
+    val suite = TestSuite()
+    val classLoader = Thread.currentThread().contextClassLoader as URLClassLoader
+
+    // The first entry on the classpath contains the srcs from java_test
+    val classesInJar = findClassesInJar(File(classLoader.urLs[0].path))
```

Comment:
Java can run class files directly, in which case there won't be `.jar` file I'd assume, but I guess since we build and run with Bazel it's not going to be an issue

Reply:
Is there an alternative here which we should explore?

Reply:
I mean, it's kind of bad that we rely on particular index like `0`, ideally you'd scan entire classpath (can be a bit slow), so that Bazel update doesn't break your tests (they'll prob silently finish in this case which is worst thing for tests)

Reply:
With the current set up we have the following items in the `classLoader.urLs`:
```
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/execroot/__main__/bazel-out/darwin-fastbuild/bin/library/kotlin/test/sample_test.jar
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/execroot/__main__/bazel-out/darwin-fastbuild/bin/bazel/envoy_mobile_test_suite.jar
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/external/maven/v1/https/repo1.maven.org/maven2/junit/junit/4.12/junit-4.12.jar
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/external/maven/v1/https/repo1.maven.org/maven2/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/external/com_github_jetbrains_kotlin/lib/kotlin-stdlib.jar
/private/var/tmp/_bazel_achiu/e95f0937739315919b86dd2afc724e3e/external/maven/v1/https/repo1.maven.org/maven2/org/assertj/assertj-core/3.9.0/assertj-core-3.9.0.jar
/private/var/tmp/_bazel_achiu/install/a134bf748a9bb448b9801b24812c57ec/_embedded_binaries/embedded_tools/tools/jdk/TestRunner_deploy.jar
```

I'm not sure what a more reliable heuristic would be to find the test class files. Should we instead just unpack all the `classLoader.urLs` jars and look for classes which have the name `*Test`?

Reply:
Another option is for us to suffix the bazel rules with `_test` but it'd be confusing for the person defining a test target

Reply:
I tried running a look up of all the jars in the context. It took ~300s to run a single test target (versus 0.3s without the exhaustive look up).

I don't think this is a reasonable solution to ensure we are not breaking on bazel updates since we would incur an incredible amount of development productivity hit here.

I've updated the PR to exclude non-envoy packages (`io.envoyproxy.envoymobile`). The bazel tests will fail if there are no tests in the test target.

This _should_ cover the case where the first `urL` is not the test target since we would fail if there are no tests within the envoy package space.